gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/Auth.java

```diff
@@ -18,33 +18,258 @@
+        return new Sigv4(regionName, awsCredentialsProvider, serviceName);
+    }
+
     public static class Basic extends Auth {
 
         private final String username;
         private final String password;
 
-        private Basic(final String username, final String password ) {
+        private Basic(final String username, final String password) {
             this.username = username;
             this.password = password;
         }
 
         @Override
-        public FullHttpRequest apply(FullHttpRequest fullHttpRequest) {
+        public FullHttpRequest apply(final FullHttpRequest fullHttpRequest) {
             final String valueToEncode = username + ":" + password;
             fullHttpRequest.headers().add(HttpHeaderNames.AUTHORIZATION,
                     "Basic " + Base64.getEncoder().encodeToString(valueToEncode.getBytes()));
             return fullHttpRequest;
         }
     }
+
+    public static class Sigv4 extends Auth {
+        private final AWSCredentialsProvider awsCredentialsProvider;
+        private final AWS4Signer aws4Signer;
+
+        private Sigv4(final String regionName, final AWSCredentialsProvider awsCredentialsProvider, final String serviceName) {
+            this.awsCredentialsProvider = awsCredentialsProvider;
+
+            aws4Signer = new AWS4Signer();
+            aws4Signer.setRegionName(regionName);
+            aws4Signer.setServiceName(serviceName);
+        }
+
+        @Override
+        public FullHttpRequest apply(final FullHttpRequest fullHttpRequest) {
+            try {
+                // Convert Http request into an AWS SDK signable request
+                final SignableRequest<?> awsSignableRequest = toSignableRequest(fullHttpRequest);
+
+                // Sign the AWS SDK signable request (which internally adds some HTTP headers)
+                final AWSCredentials credentials = awsCredentialsProvider.getCredentials();
+                aws4Signer.sign(awsSignableRequest, credentials);
+
+                // extract session token if temporary credentials are provided
+                String sessionToken = "";
+                if ((credentials instanceof BasicSessionCredentials)) {
+                    sessionToken = ((BasicSessionCredentials) credentials).getSessionToken();
+                }
+
+                // todo: confirm is needed to replace header `Host` with `host`
+                fullHttpRequest.headers().remove(HttpHeaderNames.HOST);
+                fullHttpRequest.headers().add(HOST, awsSignableRequest.getHeaders().get(HOST));
+                fullHttpRequest.headers().add(X_AMZ_DATE, awsSignableRequest.getHeaders().get(X_AMZ_DATE));
+                fullHttpRequest.headers().add(AUTHORIZATION, awsSignableRequest.getHeaders().get(AUTHORIZATION));
+
+                if (!sessionToken.isEmpty()) {
+                    fullHttpRequest.headers().add(X_AMZ_SECURITY_TOKEN, sessionToken);
+                }
+            } catch (final Throwable t) {
```

Comment:
We probably need slightly better handling of errors here rather than just RuntimeException.